### DISCUSSION TITLE
Communicating with outer functions

### DISCUSSION THREAD

**Discussion by ulzu** [fish-shell/fish-shell#10264] [Q&A] [answered]
How come the only way to communicate with outer functions is through globals or the kludgy echo-as-return hack? Why do we have no way to re-export variables back to the outer function?

**Comment by faho** [comment]
Because commands also couldn't.

Fish's model is fairly simple and like a classic shell, if you want a shell that has more structured ways of communication others are available.

**Reply by faho** [reply, marked as answer]
That being said, #3540 is still open.

**Reply by ulzu** [reply]
Noo, I like fish a lot. And consistency is good, I'm not at all expressing a criticism.

As functions run in the same subprocess, I considered them to just be a list of commands, with the additional scoping. Just like a scope introduced by an if statement. I was confused that it is not quite it. This was compounded by how `-f` flag allowed variables to go upscope. I first understood this as setting a variable within a sub-environment (sub-global environment?) available only for the items within this function.
And `-x` flag is defined as making a variable be available to child *processes*, while with a function we remain in the same process, so I was imagining that "function-scope" also extends to functions called therein – a veritable sub-environment. That's what brought me here with the question.

No demands here :)
